src/dir/glob.cr

```diff
@@ -1,266 +1,369 @@
+  # ditto
+  def self.glob(patterns : Enumerable(String), **options, &block : String -> _)
+    Globber.new(**options).glob(patterns) do |path|
+      yield path
+    end
+  end
 
-      if starts_with_path_separator
-        dir = root
-      else
-        dir = cwd
+  # :nodoc:
+  struct Globber
+    record DirectoriesOnly
+    record ConstantEntry, path : String
+    record EntryMatch, pattern : String do
+      def matches?(string)
+        File.fnmatch(pattern, string)
       end
-
-      if ptrn.includes? "**"
-        recursion_depth = Int32::MAX
+    end
+    record StartRecursiveDirectories
+    record RecursiveDirectories
+    record ConstantDirectory, path : String
+    record RootDirectory
+    record DirectoryMatch, pattern : String do
+      def matches?(string)
+        File.fnmatch(pattern, string)
       end
+    end
+    alias PatternType = DirectoriesOnly | ConstantEntry | EntryMatch | StartRecursiveDirectories | RecursiveDirectories | ConstantDirectory | RootDirectory | DirectoryMatch
 
-      # optimize the glob by starting with the directory
-      # which is as nested as possible:
-      lastidx = 0
-      depth = 0
-      escaped = false
-      last_is_file_separator = false
-      count = 0
-      nested_path = String.build do |str|
-        ptrn.each_char_with_index do |c, i|
-          if c == '\\'
-            escaped = true
-            last_is_file_separator = false
-            next
-          elsif c == File::SEPARATOR
-            unless last_is_file_separator
-              depth += 1
-              str << c
-              count += 1
-            end
-            lastidx = count
-            last_is_file_separator = true
-          elsif !escaped && special.includes? c
-            break
-          else
-            last_is_file_separator = false
-            str << c
-            count += 1
+    property? allow_dots
+
+    def initialize(@allow_dots = false)
+    end
+
+    def glob(patterns : Enumerable(String), &block : String -> _)
+      patterns.each do |pattern|
+        sequences = compile(pattern)
+
+        sequences.each do |sequence|
+          run(sequence) do |match|
+            yield match
           end
-          escaped = false
         end
       end
-      nested_path = nested_path[0...lastidx]
+    end
 
-      recursion_depth -= depth if recursion_depth != Int32::MAX
-      dir = File.join(dir, nested_path)
-      if !nested_path.empty? && nested_path[0] == File::SEPARATOR
-        nested_path = nested_path[1..-1]
-      end
+    private def compile(pattern, patterns = [] of Array(PatternType))
+      reader = Char::Reader.new(pattern)
 
-      regex = glob2regex(ptrn)
+      lbrace = nil
+      rbrace = nil
+      alt_start = nil
 
-      scandir(dir, nested_path, regex, 0, recursion_depth, wants_dir) do |path|
-        if starts_with_path_separator
-          yield File::SEPARATOR + path
-        else
-          yield path
+      alternatives = [] of String
```

Comment:
Maybe lazily create this array?

Reply:
I'm not sure I understand where you're going...
Do you mean something like:
```crystal
alternatives : Array(String)?
# ...
(alternatives ||= [] of String) << pattern.byte_slice(start, reader.pos - start)
# ...
if alternatives
```

Reply:
Yes, like that. I don't know if it affects performance, though.

Reply:
I would rather leave it as is. If it has any impact at all it is totally insignificant. This method is only run once for each glob pattern as long as it doesn't have any curly braces - if it does, this might actually make it worse because of the additional nil checks.